README.md
# common-less-replacer README

This is the README for your extension "common-less-replacer". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: Enable/disable this extension.
* `myExtension.thing`: Set to `blah` to do something.

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

---

## Following extension guidelines

Ensure that you've read through the extensions guidelines and follow the best practices for creating your extension.

* [Extension Guidelines](https://code.visualstudio.com/api/references/extension-guidelines)

## Working with Markdown

You can author your README using Visual Studio Code. Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux).
* Toggle preview (`Shift+Cmd+V` on macOS or `Shift+Ctrl+V` on Windows and Linux).
* Press `Ctrl+Space` (Windows, Linux, macOS) to see a list of Markdown snippets.

## For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "common-less-replacer" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/LessVarCompletionProvider.ts
/*
 * @Author: guowei26
 * @Date: 2023-04-24 19:03:53
 * @LastEditors: guowei26
 * @LastEditTime: 2023-04-27 17:20:58
 * @FilePath: /common-less-replacer/src/LessVarCompletionProvider.ts
 * @Description: less输入文字监测
 */
import * as vscode from 'vscode';
import {LessAstListIProps, LessTypeEnum} from './type';

const createCompletion = (list: Array<any>, character?: string) => {
    return list.map(item => {
        const currentCompletionItem = new vscode.CompletionItem(
            `${item.key} : ${item.value}`,
            vscode.CompletionItemKind.Variable
        );
        currentCompletionItem.detail = item.key;
        currentCompletionItem.insertText = item.key;

        return currentCompletionItem;
    });
};

export default class ColorCompletionProvider implements vscode.CompletionItemProvider {
    private lessAst: LessAstListIProps;
    private character?: string;

    provideCompletionItems(document: vscode.TextDocument, position: vscode.Position) {
        const line = document.lineAt(position.line).text;
        const currentLessAst = this.lessAst.reduce((total, current) => {
            return {...total, ...current};
        }, {});
        // 存在特定字符触发
        if (this.character) {
            if (!line.endsWith(this.character)) {
                return [];
            }
            const realLinePrefix = line.replace(/;/g, '').split(':');
            if (realLinePrefix.length === 2) {
                const value = realLinePrefix[1].replace(/\s*/g, '');
                const currentLessAst = this.lessAst.reduce((total, current) => {
                    return {...total, ...current};
                }, {});
                const targetVar: Array<{key: string; value: string}> = [];
                Object.keys(currentLessAst).forEach((key: string) => {
                    const currentLess = currentLessAst[key];
                    if (currentLess.value.includes(value)) {
                        targetVar.push({key, value: currentLess.value});
                    }
                });
                if (targetVar.length > 0) {
                    const currentCompletionList: vscode.CompletionItem[] = createCompletion(targetVar, this.character);
                    return currentCompletionList;
                }
                return [];
            }
            return [];
        }
        const targetVar: Array<{key: string; value: string; type: LessTypeEnum}> = [];
        Object.keys(currentLessAst).forEach((key: string) => {
            const currentLess = currentLessAst[key];
            targetVar.push({key, value: currentLess.value, type: currentLess.type});
        });
        const currentCompletionList: vscode.CompletionItem[] = createCompletion(targetVar);
        return currentCompletionList;
    }

    constructor(lessAst: LessAstListIProps, character?: string) {
        this.lessAst = lessAst;
        this.character = character;
    }
}


src/extension.ts
/*
 * @Author: guowei26
 * @Date: 2023-04-21 14:53:27
 * @LastEditors: guowei26
 * @LastEditTime: 2023-04-27 16:39:19
 * @FilePath: /common-less-replacer/src/extension.ts
 */
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import * as util from 'util';
import * as less from 'less';
import * as events from 'events';
import LessVarCompletionProvider from './LessVarCompletionProvider';
import {LessTypeEnum, LineAstItemIProps, LessAstListIProps} from './type';

// AST to Obj
const transferLessAstToLessObj = (rules: any[]) => {
    const lessObj: Record<string, LineAstItemIProps> = {};
    // TODO 暂未处理less变量
    rules.forEach(item => {
        const value = item.value?.value;
        if (item.type === LessTypeEnum.DECLARATION) {
            if (typeof value === 'string') {
                lessObj[item.name] = {
                    value: value,
                    type: LessTypeEnum.NUMBER
                };
            }
            if (Array.isArray(value) && value.length === 1) {
                const currentValue = value[0]?.value;
                if (
                    Array.isArray(currentValue) &&
                    currentValue.length === 1 &&
                    currentValue[0].type === LessTypeEnum.COLOR
                ) {
                    lessObj[item.name] = {
                        value: currentValue[0].value,
                        type: LessTypeEnum.COLOR
                    };
                }
            }
        }
    });
    return lessObj;
};

export function activate(context: vscode.ExtensionContext) {
    const config = vscode.workspace.getConfiguration();
    const filePathList = config.get<string[]>('conf.less.path') || [];
    const lessAst: LessAstListIProps = [];
    const fileWather: vscode.FileSystemWatcher[] = [];

    // 保存用户输入的less路径
    const disposable = vscode.commands.registerCommand('common-less-replacer.helloWorld', async () => {
        const value = await vscode.window.showInputBox({
            prompt: '请输入所配置的less路径，多个的话请以逗号隔开',
            value: filePathList?.join(',')
        });
        if (value) {
            const pathList = value.split(',').filter(item => !!item);
            await config.update('conf.less.path', pathList, vscode.ConfigurationTarget.Workspace, true);
        }
    });

    vscode.workspace.onDidChangeTextDocument(event => {
        const editor = vscode.window.activeTextEditor;
        if (!editor) {
            return;
        }
        if (event.contentChanges.length) {
            const change = event.contentChanges[0];
            const line = change.range.start.line;
            const character = change.range.start.character;

            if (change.text === '' && character > 0) {
                // 删除了字符，重新启用自动完成功能
                const newPosition = new vscode.Position(line, character - 1);
                vscode.commands.executeCommand('vscode.executeCompletionItemProvider', event.document.uri, newPosition);
            }
        }
    });

    // less输入时自动补全-模糊搜索补全
    const autoLessProvider = vscode.languages.registerCompletionItemProvider(
        'less',
        new LessVarCompletionProvider(lessAst)
    );

    // less输入后,;结尾触发自动补全
    const rightLessProvider = vscode.languages.registerCompletionItemProvider(
        'less',
        new LessVarCompletionProvider(lessAst, ';'),
        ';'
    );

    // 根据用户输入的less文件路径，创建文件修改监听
    filePathList.forEach((url, urlIndex) => {
        const fileUrl = vscode.workspace.rootPath && path.join(vscode.workspace.rootPath, url);
        const existFiles = fileUrl && fs.existsSync(fileUrl);
        if (existFiles) {
            const watcher = vscode.workspace.createFileSystemWatcher(fileUrl);
            fileWather.push(watcher);
            const fileContent = fs.readFileSync(fileUrl, 'utf-8');
            less.default.parse(fileContent, {processImports: false}, function (error: any, result: any) {
                if (error) {
                    return;
                }
                const lessObj = transferLessAstToLessObj(result.rules);
                if (lessAst?.[urlIndex]) {
                    lessAst[urlIndex] = lessObj;
                } else {
                    lessAst.push(lessObj);
                }
            });
            console.log(lessAst);
        }
    });

    fileWather.forEach((watcher, watchIndex) => {
        context.subscriptions.push(watcher);
        watcher.onDidChange(async event => {
            if (event.fsPath) {
                const fileContent = await vscode.workspace.fs.readFile(event);
                const decorder = new util.TextDecoder();
                const content = decorder.decode(fileContent);
                // TODO 支持 @import引入的less文件
                // TODO 删除less文件后的处理
                const result = await less.default.parse(content, {processImports: false});
                const lessObj = transferLessAstToLessObj(result.rules);
                if (lessAst?.[watchIndex]) {
                    lessAst[watchIndex] = lessObj;
                } else {
                    lessAst.push(lessObj);
                }
                console.log('改变后的less内容', lessAst);
            }
        });
    });
    context.subscriptions.push(autoLessProvider, rightLessProvider, disposable);
}

export function deactivate() {
    // TODO wather删除
}
